### PR TITLE
DRIVERS-2651 Add decimal128 clamped zeros tests with very large exponents.

### PR DESCRIPTION
**Pull request by matthewdale**
[DRIVERS-2651](https://jira.mongodb.org/browse/DRIVERS-2651)

Add decimal128 Extended JSON parse tests for clamped zeros with very large positive and negative exponents to catch any issues in iterative clamping logic.

Please complete the following before merging:
- [x] Update changelog.
- [x] Make sure there are generated JSON files from the YAML test files.
- [ ] Test changes in at least one language driver.
- [ ] Test these changes against all server versions and topologies (including standalone, replica set, sharded clusters, and serverless).

### CODE REVIEW

**Comment by matthewdale** [comment]
@jyemin that's really interesting! We could add these cases as expected parse errors instead of valid "degenerate_extjson". I'll check if there's an existing trend among drivers.

**Comment by jmikola** [comment]
Created https://github.com/mongodb/mongo-php-driver/pull/1433 to POC this using the PHP driver, which effectively tests libbson from the C driver.

There's no infinite loop (as mentioned in [DRIVERS-2651](https://jira.mongodb.org/browse/DRIVERS-2651)), but I do see failures converting the degenerate extJSON to Canonical BSON. The PHP corpus tests exercise two code paths (one uses a BSON object as an intermediary representation) but this is really just one failure. 

In the diffs below, the `-` segment is expected output and `+` is actual.

----

```json
{
    "description": "clamped zeros with a large negative exponent",
    "canonical_bson": "180000001364000000000000000000000000000000000000",
    "degenerate_extjson": "{\"d\" : {\"$numberDecimal\" : \"0E-999999999999\"}}",
    "canonical_extjson": "{\"d\" : {\"$numberDecimal\" : \"0E-6176\"}}"
}
```

```diff
 // Degenerate extJSON -> Canonical BSON
-180000001364000000000000000000000000000000000000
+180000001364000000000000000000000000000000fe5f00
 // Degenerate extJSON -> BSON object -> Canonical BSON
-180000001364000000000000000000000000000000000000
+180000001364000000000000000000000000000000fe5f00
```

----

```json
{
    "description": "clamped zeros with a large positive exponent",
    "canonical_bson": "180000001364000000000000000000000000000000FE5F00",
    "degenerate_extjson": "{\"d\" : {\"$numberDecimal\" : \"0E+999999999999\"}}",
    "canonical_extjson": "{\"d\" : {\"$numberDecimal\" : \"0E+6111\"}}"
}
```

```diff
 // Degenerate extJSON -> Canonical BSON
-180000001364000000000000000000000000000000fe5f00
+180000001364000000000000000000000000000000000000
 // Degenerate extJSON -> BSON object -> Canonical BSON
-180000001364000000000000000000000000000000fe5f00
+180000001364000000000000000000000000000000000000
```

----

```json
{
    "description": "clamped negative zeros with a large negative exponent",
    "canonical_bson": "180000001364000000000000000000000000000000008000",
    "degenerate_extjson": "{\"d\" : {\"$numberDecimal\" : \"-0E-999999999999\"}}",
    "canonical_extjson": "{\"d\" : {\"$numberDecimal\" : \"-0E-6176\"}}"
}
```

```diff
 // Degenerate extJSON -> Canonical BSON
-180000001364000000000000000000000000000000008000
+180000001364000000000000000000000000000000fedf00
 // Degenerate extJSON -> BSON object -> Canonical BSON
-180000001364000000000000000000000000000000008000
+180000001364000000000000000000000000000000fedf00
```

----

```json
{
    "description": "clamped negative zeros with a large positive exponent",
    "canonical_bson": "180000001364000000000000000000000000000000FEDF00",
    "degenerate_extjson": "{\"d\" : {\"$numberDecimal\" : \"-0E+999999999999\"}}",
    "canonical_extjson": "{\"d\" : {\"$numberDecimal\" : \"-0E+6111\"}}"
}
```

```diff
 // Degenerate extJSON -> Canonical BSON
-180000001364000000000000000000000000000000fedf00
+180000001364000000000000000000000000000000008000
 // Degenerate extJSON -> BSON object -> Canonical BSON
-180000001364000000000000000000000000000000fedf00
+180000001364000000000000000000000000000000008000
```

File: source/bson-corpus/tests/decimal128-5.json [hunk 1]
```diff
@@ -151,6 +151,12 @@
             "degenerate_extjson": "{\"d\" : {\"$numberDecimal\" : \"0E-6177\"}}",
             "canonical_extjson": "{\"d\" : {\"$numberDecimal\" : \"0E-6176\"}}"
         },
+        {
+            "description": "clamped zeros with a large negative exponent",
+            "canonical_bson": "180000001364000000000000000000000000000000000000",
+            "degenerate_extjson": "{\"d\" : {\"$numberDecimal\" : \"0E-999999999999\"}}",
+            "canonical_extjson": "{\"d\" : {\"$numberDecimal\" : \"0E-6176\"}}"
+        },
```
Comment (jmikola) [inline source/bson-corpus/tests/decimal128-5.json:159]: Was the reason for adding these tests to `decimal128-5.json` because they copy the preceding test case and just modify `degenerate_extjson`?

The tests in this file are taken from `dqBase.decTest` in `decTest.zip` from https://speleotrove.com/decimal/dectest.html (as referenced in the [Decimal128 spec](https://github.com/mongodb/specifications/blob/master/source/bson-decimal128/decimal128.rst#tests)).

```
-- clamped zeros [see also clamp.decTest]
dqbas1075 toSci   0e+10000  ->  0E+6111 Clamped
dqbas1076 toSci   0e-10000  ->  0E-6176  Clamped
dqbas1077 toSci  -0e+10000  -> -0E+6111 Clamped
dqbas1078 toSci  -0e-10000  -> -0E-6176  Clamped
```

I think there's value in keeping these "ported" files as-is and adding these tests either a new file or the end of an existing one. `decimal128-1.json` and `decimal128-6.json` look to be organic tests not ported from the aforementioned test suite.

That said, I won't protest if everyone else prefers keeping them here.

Comment (matthewdale) [reply source/bson-corpus/tests/decimal128-5.json:159]: There wasn't a clear pattern for where to add new `decimal128` tests, so I just added them near existing similar tests. However, I can definitely add them to either `decimal128-1.json` or `decimal128-6.json` instead.

**Comment by matthewdale** [comment]
@jyemin doing some quick investigation into that exception message, it seems possible that the Java BigDecimal parser will throw an exception if the exponent has more than 10 digits. I've updated the tests to use a number with exactly 10 digits, which should hopefully pass the Java tests now.

**Review by jyemin — COMMENTED** [review]
Now I'm getting:

```
Caused by: java.lang.NumberFormatException: Exponent overflow.
	at java.base/java.math.BigDecimal.<init>(BigDecimal.java:583)
	at java.base/java.math.BigDecimal.<init>(BigDecimal.java:471)
	at java.base/java.math.BigDecimal.<init>(BigDecimal.java:900)
	at org.bson.types.Decimal128.parse(Decimal128.java:131)
	at org.bson.json.JsonReader.visitNumberDecimalExtendedJson(JsonReader.java:1290)
	... 69 more
```

The BigDecimal code is:

```
                        if ((int) exp != exp) // overflow
                            throw new NumberFormatException("Exponent overflow.");
```

and `exp = 9999999999 (0x2540BE3FF)`

**Comment by matthewdale** [comment]
@jyemin Hah, more validation! I've updated the test cases to use max/min int32 values for the exponents. It still triggers the problem in the old Go driver and hopefully passes the Java tests now.

**Review by jyemin — COMMENTED** [review]
Down from four to two failures now.  They are similar, so here's the first:

```
Exception converting value '0E-2147483648' to type org.bson.types.Decimal128
org.bson.json.JsonParseException: Exception converting value '0E-2147483648' to type org.bson.types.Decimal128
	at app//org.bson.json.JsonReader.visitNumberDecimalExtendedJson(JsonReader.java:1292)
	at app//org.bson.json.JsonReader.visitExtendedJSON(JsonReader.java:670)
	at app//org.bson.json.JsonReader.readBsonType(JsonReader.java:166)
	at app//org.bson.codecs.BsonDocumentCodec.decode(BsonDocumentCodec.java:85)
	at app//org.bson.BsonDocument.parse(BsonDocument.java:66)
	at app//org.bson.GenericBsonTest.runValid(GenericBsonTest.java:144)
	at app//org.bson.GenericBsonTest.shouldPassAllOutcomes(GenericBsonTest.java:83)
Caused by: java.lang.NumberFormatException: Scale out of range.
	at java.base/java.math.BigDecimal.adjustScale(BigDecimal.java:705)
	at java.base/java.math.BigDecimal.<init>(BigDecimal.java:595)
	at java.base/java.math.BigDecimal.<init>(BigDecimal.java:471)
	at java.base/java.math.BigDecimal.<init>(BigDecimal.java:900)
	at org.bson.types.Decimal128.parse(Decimal128.java:131)
	at org.bson.json.JsonReader.visitNumberDecimalExtendedJson(JsonReader.java:1290)
	... 69 more
```

The check is:

```
    private int adjustScale(int scl, long exp) {
        long adjustedScale = scl - exp;
        if (adjustedScale > Integer.MAX_VALUE || adjustedScale < Integer.MIN_VALUE)
            throw new NumberFormatException("Scale out of range.");
        scl = (int) adjustedScale;
        return scl;
    }
```

When the exception is thrown:

```
scl = 0
exp = -2147483648 (0xFFFFFFFF80000000)
adjustedScale = 2147483648 (0x80000000)
```

**Comment by matthewdale** [comment]
@jyemin good suggestion. I adjusted the negative exponent by +1.

**Review by jyemin — APPROVED** [review]
All green now for Java driver.

LGTM

**Review by blink1073 — APPROVED** [review]
LGTM!

**Comment by jmikola** [comment]
From https://github.com/mongodb/specifications/pull/1432#issuecomment-1593893814

> I've updated the tests to use a number with exactly 10 digits, which should hopefully pass the Java tests now.

From https://github.com/mongodb/specifications/pull/1432#issuecomment-1594012348

> I've updated the test cases to use max/min int32 values for the exponents. It still triggers the problem in the old Go driver and hopefully passes the Java tests now.

@matthewdale: It looks like these changes resolved the [previous failures](https://github.com/mongodb/specifications/pull/1432#issuecomment-1592128172) I observed in libbson (via PHPC).
